grpcbp/server_middlewares_test.go

```diff
@@ -178,6 +183,261 @@ func (t *mockService) PingEmpty(ctx context.Context, req *pb.Empty) (*pb.PingRes
+
+	testCases := []struct {
+		name    string
+		wantErr codes.Code
+		code    string
+		success string
+		method  string
+	}{
+		{
+			name:    "success",
+			wantErr: codes.OK,
+			code:    "OK",
+			success: "true",
+			method:  "PingStream",
+		},
+		{
+			name:    "err",
+			wantErr: codes.NotFound,
+			code:    "NotFound",
+			success: "false",
+			method:  "PingStream",
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			serverLatencyDistribution.Reset()
+			serverRPCRequestCounter.Reset()
+			serverActiveRequests.Reset()
+			clientLatencyDistribution.Reset()
+			clientRPCRequestCounter.Reset()
+			clientActiveRequests.Reset()
+
+			if tt.success == "false" {
+				service.err = status.Errorf(tt.wantErr, "test err: %v", tt.wantErr.String())
+			}
+
+			serverLabelValues := []string{
+				serviceName,
+				tt.method,
+				serverStream,
+				tt.success,
+				tt.code,
+			}
+
+			serverRequestsLabelValues := []string{
+				serviceName,
+				tt.method,
+			}
+
+			clientLabelValues := []string{
+				serviceName,
+				"/mwitkow.testproto.TestService/" + tt.method,
+				clientStream,
+				tt.success,
+				tt.code,
+				serverName,
+			}
+
+			clientRequestsLabelValues := []string{
+				serviceName,
+				tt.method,
+				serverName,
+			}
+
+			defer prometheusbp.MetricTest(t, "server latency", serverLatencyDistribution).CheckExists()
+			defer prometheusbp.MetricTest(t, "server rpc count", serverRPCRequestCounter, serverLabelValues...).CheckDelta(1)
+			defer prometheusbp.MetricTest(t, "server active requests", serverActiveRequests, serverRequestsLabelValues...).CheckDelta(0)
+			defer prometheusbp.MetricTest(t, "client latency", clientLatencyDistribution).CheckExists()
+			defer prometheusbp.MetricTest(t, "client rpc count", clientRPCRequestCounter, clientLabelValues...).CheckDelta(1)
+			defer prometheusbp.MetricTest(t, "client active requests", clientActiveRequests, clientRequestsLabelValues...).CheckDelta(0)
+
+			ctx := context.Background()
+			clientStream, err := client.PingStream(ctx)
+			if err != nil {
+				t.Fatalf("PingStream: %v", err)
+			}
+
+			var count int
+			for {
+				switch {
+				case count < 2:
+					err = clientStream.Send(&pb.PingRequest{})
+				default:
+					err = clientStream.CloseSend()
+				}
+				if err != nil && err != io.EOF {
+					t.Fatalf("clientStream Send or CloseSend: %v", err)
+				}
+				if err == io.EOF {
+					break
+				}
+
+				_, err := clientStream.Recv()
+				if err == io.EOF {
+					break
+				}
+				if tt.success == "false" {
+					if err != nil {
+						status, _ := status.FromError(err)
+						if want, got := tt.wantErr.String(), status.Code().String(); want != got {
+							t.Fatalf("error mismatch: want %v, got %v", want, got)
+						}
+					}
+					if err == nil {
+						t.Fatal("clientStream Recv: expected err got nil")
+					}
+				}
+
+				if err != nil && tt.success == "true" {
+					t.Fatalf("clientStream Recv: %v", err)
+				}
```

Comment:
minor suggestion for improving clarity... I think you can structure it into two conditionals instead of having to rely on nesting:

```
_, err := ...
// ensure that tt.success is set correctly
if got, want := tt.success, fmt.Sprint(err == nil || err == io.EOF); got != want {
  t.Errorf("tt.success = %q, want %q (err: %v)")
}
// handle the actual error
if err != nil {
  // ...
  break
}
count++

```